1-Draft/RFCxxxx-Ternary-Operator.md

```diff
@@ -0,0 +1,154 @@
+In those cases, we will force to start a new token upon seeing characters `?` and `:` when scanning for a number,
+so that that number constant expressions can work with ternary operators more intuitively.
+
+Note that, the handling of `?` and `:` doesn't change how those two characters can be used in variable expressions,
+and also doesn't affect function and commands with those two characters in names.
+
+#### AST Visitors
+
+Given the fact that we are adding a new AST type, necessary changes are required to `ICustomAstVisitor2`, `AstVisitor2`,
+and all other visitor types that implemnent those two.
+Obvious examples are the `VariableAnalysis` and `Compiler` as you have to update them properly to correctly enable the new language element.
+The not-so-obvious ones include `ConstantValueVisitor`, `SafeValueVisitor`, `TypeInferenceVisitor` and more.
+Take the `TypeInferenceVisitor` as an instance, failing to update it may cause tab completion issues that are hard to catch.
+
+It's worth to discuss the `InternalVisit` implemention of the new AST type a bit.
+Many `Find*Visitor` types depend on this method to search through an AST tree,
+and almost all of them derive from `AstVisitor` instead of `AstVisitor2` because
+currently the AST types covered in `AstVisitor2` all have their own special scopes,
+such as the `TypeDefinitionAst` and `DynamicKeywordStatement`.
+In order to avoid making changes to all those existing `Find*Visitor` types,
+the `InternalVisit` method of `TernaryExpressionAst` is implemented as follows to accept an `AstVisitor` visitor for its AST members.
+In this way, those existing `Find*Visitor` types can work with this new AST type to search through its members.
+
+```c#
+// This pattern appears in several other ASTs.
+internal override AstVisitAction InternalVisit(AstVisitor visitor)
+{
+    var action = AstVisitAction.Continue;
+    if (visitor is AstVisitor2 visitor2)
+    {
+        action = visitor2.VisitTernaryExpression(this);
+        if (action == AstVisitAction.SkipChildren)
+            return visitor.CheckForPostAction(this, AstVisitAction.Continue);
+    }
+
+    if (action == AstVisitAction.Continue)
+        action = Condition.InternalVisit(visitor);
+
+    if (action == AstVisitAction.Continue)
+        action = IfTrue.InternalVisit(visitor);
+
+    if (action == AstVisitAction.Continue)
+        action = IfFalse.InternalVisit(visitor);
+
+    return visitor.CheckForPostAction(this, action);
+}
+```
+
+### Alternate Proposals and Considerations
+
+Different syntax proposals were brought up in the discussions:
+
+```none
+## Use '-then -else' instead of '? :'
+<condition-expression> -then <then-expression> -else <else-expression>
+```
+```none
+## Use syntax similar to '-replace' and '-f'
+<condition-expression> -ternary <then-expression>, <else-expression>
+```
+
+After a discussion at length, the PowerShell committee agrees to align with the C# syntax given
```

Comment:
We should add that one of the concerns with using `?` is the visual acuity impact of scanning code where `?` typically has only been used as the alias for `Where-Object`.  However, we expect that as we implement other language features like null-coalescing, null-conditional assignment, and null-conditional access, we are likely to use `?` in some form.  Also, users can choose to opt out of using this new language feature.